File: FrontEnd/src/lib/walletService.ts
import { apiClient } from './apiClient';

export interface WalletDTO {
  id: string;
  userId: string;
  walletName: string;
  balance: number;
  currency: string;
  isActive: boolean;
  isDemoFunded: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface WalletResponse {
  status: 'success' | 'error';
  message?: string;
  ref?: string;
  newBalance?: number;
  receiverBalance?: number;
  wallet?: WalletDTO;
}

class WalletService {
  async getActiveWallet(): Promise<WalletDTO> {
    console.log('[Wallet] Fetching active wallet');

    const response = await apiClient.get<WalletDTO>('/wallets/active');

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch active wallet');
    }

    console.log('[Wallet] Active wallet balance:', response.data.balance);
    return response.data;
  }

  async getWalletById(walletId: string): Promise<WalletDTO> {
    console.log('[Wallet] Fetching wallet by ID', { walletId });

    const response = await apiClient.get<WalletDTO>(`/wallets/${walletId}`);

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch wallet');
    }

    return response.data;
  }

  async getUserWallets(): Promise<WalletDTO[]> {
    console.log('[Wallet] Fetching all user wallets');

    const response = await apiClient.get<WalletDTO[]>('/wallets/list');

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fetch wallets');
    }

    return response.data;
  }

  async fundDemoWallet(amount: number): Promise<WalletDTO> {
    console.log('[Wallet] Funding demo wallet', { amount });

    const response = await apiClient.post<WalletDTO>(`/wallets/demo-fund?amount=${amount}`);

    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to fund demo wallet');
    }

    console.log('[Wallet] Demo wallet funded, new balance:', response.data.balance);
    return response.data;
  }

  async getWalletSnapshot(userId: string): Promise<WalletResponse> {
    try {
      const wallet = await this.getActiveWallet();
      return {
        status: 'success',
        wallet: wallet,
        newBalance: wallet.balance,
      };
    } catch (error) {
      console.error('[Wallet] Failed to get wallet snapshot', error);
      return {
        status: 'error',
        message: error instanceof Error ? error.message : 'Failed to fetch wallet',
      };
    }
  }

  formatKES(amount: number): string {
    return `Ksh${amount.toLocaleString('en-KE', { minimumFractionDigits: 2 })}`;
  }
}

export const walletService = new WalletService();

File: Backend/EmergencyWallet/src/main/java/org/NovaGroup/EmergencyWallet/service/WalletService.java
package org.NovaGroup.EmergencyWallet.service;

import lombok.extern.slf4j.Slf4j;
import org.NovaGroup.EmergencyWallet.dto.WalletDTO;
import org.NovaGroup.EmergencyWallet.entity.User;
import org.NovaGroup.EmergencyWallet.entity.Wallet;
import org.NovaGroup.EmergencyWallet.exception.ApiException;
import org.NovaGroup.EmergencyWallet.repository.UserRepository;
import org.NovaGroup.EmergencyWallet.repository.WalletRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
public class WalletService {

    @Autowired
    private WalletRepository walletRepository;

    @Autowired
    private UserRepository userRepository;

    public WalletDTO getWalletById(String walletId) {
        log.debug("Fetching wallet: {}", walletId);
        Wallet wallet = walletRepository.findById(walletId)
            .orElseThrow(() -> new ApiException("Wallet not found", "WALLET_NOT_FOUND"));
        return mapToDTO(wallet);
    }

    public WalletDTO getUserActiveWallet(String userId) {
        log.debug("Fetching active wallet for user: {}", userId);
        Wallet wallet = walletRepository.findByUserIdAndIsActiveTrue(userId)
            .orElseThrow(() -> new ApiException("No active wallet found", "WALLET_NOT_FOUND"));
        return mapToDTO(wallet);
    }

    public List<WalletDTO> getUserWallets(String userId) {
        log.debug("Fetching all wallets for user: {}", userId);
        return walletRepository.findByUserId(userId)
            .stream()
            .map(this::mapToDTO)
            .collect(Collectors.toList());
    }

    @Transactional
    public void fundDemoWallet(String userId, BigDecimal amount) {
        log.info("Funding demo wallet for user: {} with amount: {}", userId, amount);
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ApiException("User not found", "USER_NOT_FOUND"));

        Wallet wallet = walletRepository.findByUserIdAndIsActiveTrue(userId)
            .orElseThrow(() -> new ApiException("Wallet not found", "WALLET_NOT_FOUND"));

        wallet.setBalance(wallet.getBalance().add(amount));
        wallet.setIsDemoFunded(true);
        walletRepository.save(wallet);

        log.info("Demo wallet funded successfully. New balance: {}", wallet.getBalance());
    }

    public Wallet getWalletEntity(String walletId) {
        return walletRepository.findById(walletId)
            .orElseThrow(() -> new ApiException("Wallet not found", "WALLET_NOT_FOUND"));
    }

    public Wallet getWalletEntity(String walletId, String userId) {
        return walletRepository.findByIdAndUserId(walletId, userId)
            .orElseThrow(() -> new ApiException("Wallet not found", "WALLET_NOT_FOUND"));
    }

    private WalletDTO mapToDTO(Wallet wallet) {
        return WalletDTO.builder()
            .id(wallet.getId())
            .userId(wallet.getUser().getId())
            .walletName(wallet.getWalletName())
            .balance(wallet.getBalance())
            .currency(wallet.getCurrency())
            .isActive(wallet.getIsActive())
            .isDemoFunded(wallet.getIsDemoFunded())
            .createdAt(wallet.getCreatedAt())
            .updatedAt(wallet.getUpdatedAt())
            .build();
    }
}
